From faf3a201e33ee8012c1f3e19bc249f9307924cf1 Mon Sep 17 00:00:00 2001
From: Larry Doolittle <ldoolitt@recycle.lbl.gov>
Date: Sun, 2 Mar 2008 21:14:35 +0000
Subject: [PATCH] Fix null pointer dereference

This is a real bug in upstream dash, which has existed since at
least dash-0.5.1 (July 2004).  It is a latent bug in cmdtxt()
(which I think applies only to pipes) as it handles "if" commands.
The attached patch fixes it for me.

Test case:

 dash -ic 'if true; then for x in foo; do echo bar; done; fi | cat'
---
 src/jobs.c |    5 +++--
 1 files changed, 3 insertions(+), 2 deletions(-)

diff --git a/src/jobs.c b/src/jobs.c
index 529d615..40dc8f6 100644
--- a/src/jobs.c
+++ b/src/jobs.c
@@ -1235,11 +1235,12 @@ donode:
 		cmdputs("if ");
 		cmdtxt(n->nif.test);
 		cmdputs("; then ");
-		n = n->nif.ifpart;
 		if (n->nif.elsepart) {
-			cmdtxt(n);
+			cmdtxt(n->nif.ifpart);
 			cmdputs("; else ");
 			n = n->nif.elsepart;
+		} else {
+			n = n->nif.ifpart;
 		}
 		p = "; fi";
 		goto dotail;
-- 
1.5.5

